### PR TITLE
Create `IsNaN-20` and `IsInf-20`

### PR DESCRIPTION
**Pull request by justinchuby**
### Description

Create `IsNaN-20` and `IsInf-20` to include all floating point types. Note that even though float8 types like e4m3fnuz do not have representation for infinity, we still define the operator on these types so that the operators can be conveniently applied on all floating point types. 

### Checklist 

- [x] Update the opset in `defs.cc`
- [x] Move the old definition of `old.cc`
- [x] Add the operator to `onnx/defs/operator_sets.h`
- [x] Create node tests
- [x] Disable the tests in `test_backend_onnxruntime.py`
- [x] Create reference implementation
- [x] Update version converter
- [x] Create shape inference and checker tests
- [x] Build docs

### Motivation and Context

Fixes https://github.com/onnx/onnx/issues/5260

### CODE REVIEW

**Comment by xadupre** [comment]
About isinf, e4m3fnuz has no infinite value. Should we exclude it or return false ? Same goes with e5m2fnuz.

**Comment by justinchuby** [comment]
> About isinf, e4m3fnuz has no infinite value. Should we exclude it or return false ? Same goes with e5m2fnuz.

I think we should return False. Do we know how other frameworks handle them?

**Comment by xadupre** [comment]
Only e5m2 has infinite values and CUDA uses this type to compute the gradient. The only operator available is Gemm so there is no division involved and the default behaviour is to saturate the value to the maximum. I assume nan cannot appear with this only operation.

File: onnx/backend/test/case/node/isinf.py [hunk 1]
```diff
@@ -40,4 +52,22 @@ def export_negative_infinity_only() -> None:
...
+            node,
+            inputs=[x],
+            outputs=[y],
+            name="test_isinf_negative",
+            opset_imports=[onnx.helper.make_opsetid("", 19)],
+        )
+
+    @staticmethod
+    def export_infinity_float16() -> None:
+        node = onnx.helper.make_node(
```
Comment (justinchuby) [inline onnx/backend/test/case/node/isinf.py:65]: Need float8e4m3fn instead

**Comment by gramalingam** [comment]
I agree that for the float-8 variants that do not have an infinity value, IsInf should always return false. (While this is a redundant op for this type, it may still be helpful in defining functions that work generically for all floating-types.)

File: onnx/backend/test/case/node/isinf.py (hunk 1 above)
Comment (justinchuby) [reply onnx/backend/test/case/node/isinf.py:65]: @xadupre What is the best way to create float8 tests?

**Review by gramalingam — APPROVED** [review]

**Comment by justinchuby** [comment]
I may need to exclude test_isinf_float16_cpu

**Comment by liqunfu** [comment]
> I may need to exclude test_isinf_float16_cpu

Why other tests for new and updated have been passing? And why it was passing before?

File: onnx/backend/test/case/node/isinf.py [hunk 2]
```diff
@@ -40,4 +52,22 @@ def export_negative_infinity_only() -> None:
 
         x = np.array([-1.7, np.nan, np.inf, -3.6, np.NINF, np.inf], dtype=np.float32)
         y = np.isneginf(x)
-        expect(node, inputs=[x], outputs=[y], name="test_isinf_negative")
+        expect(
+            node,
+            inputs=[x],
+            outputs=[y],
+            name="test_isinf_negative",
+            opset_imports=[onnx.helper.make_opsetid("", 19)],
```
Comment (liqunfu) [inline onnx/backend/test/case/node/isinf.py:60]: this will prevent the test case being tested with other opset versions

Comment (justinchuby) [reply onnx/backend/test/case/node/isinf.py:60]: IIRC we will use the latest opset (20) by default if not specified, which means we will disable them all in ONNX CI. Any suggestions?

cc @jcwchen

**Comment by justinchuby** [comment]
> Why other tests for new and updated have been passing? And why it was passing before?

It should have been skipped already. I am not very sure.

File: onnx/backend/test/case/node/isinf.py (hunk 2 above)
Comment (justinchuby) [reply onnx/backend/test/case/node/isinf.py]: Removed

Comment (jcwchen) [reply onnx/backend/test/case/node/isinf.py]: I think we should specify opset_version here only on demand. For example, if the old opset test failed with the latest opset schema: https://github.com/onnx/onnx/blob/bc424cd4764afc7fb8a940023cdf07b6081b02e6/onnx/backend/test/case/node/loop.py#L133.

**Comment by justinchuby** [comment]
Everything now fails though @liqunfu @jcwchen Should opset20 tests be auto skipped or do I need to manually add skips?

**Comment by jcwchen** [comment]
> Everything now fails though @liqunfu @jcwchen Should opset20 tests be auto skipped or do I need to manually add skips?

Which test failure are you referring to? The latest commit seems to complain the node backend data is inconsistent. I would suggest you recreate those node backend data from scratch (manually running command) again.

File: onnx/test/test_backend_onnxruntime.py [hunk 3]
```diff
@@ -247,6 +247,7 @@ def run_node(cls, node, inputs, device=None, outputs_info=None, **kwargs):
         "|equal"
         "|half_pixel_symmetric"
         "|identity"
+        "|isinf_float16"
```
Comment (justinchuby) [inline onnx/test/test_backend_onnxruntime.py:250]:
```suggestion
```

**Comment by justinchuby** [comment]
@jcwchen Looks like the new test is not skipped automatically? https://dev.azure.com/onnx-pipelines/onnx/_build/results?buildId=52121&view=logs&j=825fcbdb-febe-56c2-0b31-e8b200b321eb&t=46007089-1c35-5679-e243-e8ae35eaeec6&l=656

**Review by jcwchen — COMMENTED** [review]
> @jcwchen Looks like the new test is not skipped automatically? https://dev.azure.com/onnx-pipelines/onnx/_build/results?buildId=52121&view=logs&j=825fcbdb-febe-56c2-0b31-e8b200b321eb&t=46007089-1c35-5679-e243-e8ae35eaeec6&l=656

I thought https://github.com/onnx/onnx/blob/main/onnx/test/test_backend_onnxruntime.py should be covered by ORT_MAX_ONNX_OPSET_SUPPORTED_VERSION, but it seems it doesn't and we need to manually filter new versions as https://github.com/onnx/onnx/blob/main/onnx/test/test_backend_onnxruntime.py#L234-L254. It might make sense that we enable ORT_MAX_ONNX_OPSET_SUPPORTED_VERSION for test_backend_onnxruntime.py as well to save some effort. cc @xadupre for inputs. Thanks!

**Pull request merged**